ktxspec.adoc

```diff
@@ -553,35 +558,90 @@ in the reference given in the _Level Data Format_ column of
+Basis Universal is specially designed for supercompression of some
+block-compressed texture formats and includes rate distortion
+optimization.
 ====
 
 ==== Scheme Notes (Normative)
-===== Crunch CRN
-* A file that specifies Crunch CRN with base formats other than ETC,
-  ETC2 and BC[1-3] (S3TC_DXT[1-5]) must be considered invalid.
+===== Basis Universal
+* `<<_vkformat,vkFormat>>` must be `VK_FORMAT_UNDEFINED` (0x00).
+  The <<_data_format_descriptor,_Data Format Descriptor_>> must
+  indicate which color and alpha components are present by using
+  color model `KHR_DF_MODEL_RGBSDA` with zero length and size samples
+  for the present components. See <<KDF13>> Section 5.19 "Unsized
+  Formats" for details of this usage. The values of colorModel,
+  colorPrimaries, transferFunction and flags from the DFD describing
+  the pre-deflation image data must be preserved in the post-deflation
+  DFD. 
+// <<example_rg>> gives an example DFD for images that had 2
+// components prior to deflation.
+
+////
+[[example_rg]]
+.Example RG descriptor
+[width=75%]
+|================================
+32+^| *~++uint32_t++ bit~*
+^| ~31~ ^| ~30~ ^| ~29~ ^| ~28~ ^| ~27~ ^| ~26~ ^| ~25~ ^| ~24~ ^| ~23~ ^| ~22~ ^| ~21~ ^| ~20~ ^| ~19~ ^| ~18~ ^| ~17~ ^| ~16~ ^| ~15~ ^| ~14~ ^| ~13~ ^| ~12~ ^| ~11~ ^| ~10~ ^| ~9~ ^| ~8~ ^| ~7~ ^| ~6~ ^| ~5~ ^| ~4~ ^| ~3~ ^| ~2~ ^| ~1~ ^| ~0~
+32+^| *_totalSize:_* 60
+15+^| *_descriptorType:_* 0 17+^| *_vendorId:_* 0
+16+^| *_descriptorBlockSize:_* stem:[24 + (16 \times 2) = 56] 16+^| *_versionNumber:_* 2
+8+^| *_flags:_* *++ALPHA_STRAIGHT++* 8+^| *_transferFunction:_*
+*++LINEAR++* 8+^| *_colorPrimaries:_* *++BT709++* 8+^| *_colorModel:_*
+*++RGBSDA++*
+8+^| *_~texelBlockDimension3~_* 8+^| *_~texelBlockDimension2~_* 8+^| *_~texelBlockDimension1~_* 8+^| *_~texelBlockDimension0~_*
+8+^| 0 8+^| 0 8+^| 0 8+^| 0
+8+^| *_bytesPlane3:_* 0 8+^| *_bytesPlane2:_* 0 8+^| *_bytesPlane1:_* 0 8+^| *_bytesPlane0:_* 0
+8+^| *_bytesPlane7:_* 0 8+^| *_bytesPlane6:_* 0 8+^| *_bytesPlane5:_* 0 8+^| *_bytesPlane4:_* 0
+^| *_~F~_* ^| *_~S~_* ^| *_~E~_* ^| *_~L~_* 4+^| *_~channelType~_* 24+^| ~Red sample information~
+^| 0 ^| 0 ^| 0 ^| 0 4+^| *++RED++* 8+^| *_bitLength:_* 0 16+^| *_bitOffset:_* 0
+8+^| *_~samplePosition3~_* 8+^| *_~samplePosition2~_* 8+^| *_~samplePosition1~_* 8+^| *_~samplePosition0~_*
+8+^| 0 8+^| 0 8+^| 0 8+^| 0
+32+^| *_sampleLower:_* 0
+32+^| *_sampleUpper:_* 0
+^| *_~F~_* ^| *_~S~_* ^| *_~E~_* ^| *_~L~_* 4+^| *_~channelType~_* 24+^| ~Green sample information~
+^| 0 ^| 0 ^| 0 ^| 0 4+^| *++GREEN++* 8+^| *_bitLength:_* 0 16+^| *_bitOffset:_* 0
+8+^| *_~samplePosition3~_* 8+^| *_~samplePosition2~_* 8+^| *_~samplePosition1~_* 8+^| *_~samplePosition0~_*
+8+^| 0 8+^| 0 8+^| 0 8+^| 0
+32+^| *_sampleLower:_* 0
+32+^| *_sampleUpper:_* 0
+|================================
+////
+
+[NOTE]
+.Rationale
+====
+The Basis Universal encoder combines encoding to ETC1S with deflation.
+The transcoder combines inflation to ETC1S with transcoding to one
```

Comment:
KDFS has a normative definition of ETC1S (as a strict subset of ETC1), while BasisU seems to diverge from it (at least wrt endpoints encoding). In my opinion, KTX2 should not call that custom BasisU encoding as just "ETC1S".

Reply:
The definition in KDFS is supposed to be what Rich calls ETC1S. We did run the definition by Rich when we added it to KDFS. Has something changed since? If so, we need to update KDFS. There is no other use for ETC1S beyond BasisU.

Reply:
Back then (before BasisU became open-source), we expected to use ETC1S RDO with some generic lossless step. So the KDFS definition of ETC1S is correct.

The issue is that BasisU is similar to but not exactly ETC1S.

Reply:
As I pointed out, BasisU is still refers to what it is using as ETC1S. We should update KDFS as the purpose of adding ETC1S was to describe what we would be doing in KTX2.

What exactly are the differences?

Reply:
Currently, BasisU is a software product that uses "ETC1S" in its source code. Specwise, the KTX2 should refer to the BasisU bitstream (when it's available).

BasisU stores all endpoints and selectors in global tables; image slices contain only pointers to that data. Moreover, all data blocks are additionally losslessly compressed. To get GPU-ready ETC1S payload (even without transcoding), the decoder needs to decompress Huffman and to reconstruct the blocks. The last step is not documented anywhere yet.

KDFS ETC1S reference is useful for understanding the compression technique in general but certainly not enough to implement the codec.

Reply:
I think we are at cross purposes here. By referring to ETC1S I am not attempting, or claiming, to provide enough info to implement a codec. I am attempting to provide background as to why there are potentially 2 slices. I know steps are needed to reconstruct the ETC1S blocks and I know we need to refer to the BasisU bitstream, once available. There is already a place for the reference in the table of supercompression schemes.

It sounds like your earlier statement that "BasisU seems to diverge from it (at least wrt endpoints encoding)." is referring to BasisU putting the endpoints in the global table. Is that so or is there some other difference?

I will attempt rewording to make it clear the ETC1S is a building block and that further info is required to make a complete codec. I will include another T.B.D reference to the BasisU bitstream spec.

Reply:
Please take a look at 385a6b6.

Reply:
The reworded text LGTM.